File: packages/sdk-ts/src/x402/parser.ts
/**
 * Parse the body of an HTTP `402 Payment Required` response into a normalized
 * `Requirement[]`. Supports the dominant facilitators on Solana today:
 *
 *  - Coinbase  — `{ x402Version, accepts: [...] }` shape
 *  - PayAI     — Coinbase-compatible with minor field-naming variance
 *  - Corbits   — partial fixture as of 2026-05; falls through to `unknown`
 *
 * Anything that doesn't match emerges as `facilitator: 'unknown'` so the
 * caller can decide whether to surface or reject.
 */

export type Facilitator = 'coinbase' | 'payai' | 'corbits' | 'unknown';

export type Requirement = {
  facilitator: Facilitator;
  scheme: string;
  network: string;             // 'solana-mainnet' | 'solana-devnet' | 'base-mainnet' | …
  asset: string;               // SPL mint or ERC-20 address
  amount: bigint;              // smallest unit (micro-USDC for SPL USDC, wei for ERC-20)
  recipient: string;           // base58 (Solana) or hex (EVM)
  expiresAt?: Date;
  description?: string;
  /** Original entry — preserved so the encoder can echo any facilitator-specific fields. */
  raw: unknown;
};

type AcceptEntry = Record<string, unknown>;

function asString(v: unknown): string | undefined {
  return typeof v === 'string' ? v : undefined;
}

function toBigint(v: unknown): bigint | undefined {
  if (typeof v === 'bigint') return v;
  if (typeof v === 'number' && Number.isFinite(v)) return BigInt(Math.trunc(v));
  if (typeof v === 'string' && /^\d+$/.test(v)) return BigInt(v);
  return undefined;
}

/**
 * Detect the facilitator from a top-level response shape.
 *
 * Heuristics:
 *  - Coinbase emits `{ x402Version, accepts: [{ scheme, network, … }] }`.
 *  - PayAI emits the same envelope but lists `extra.facilitator === 'payai'` or
 *    a known PayAI host in `payTo`.
 *  - Corbits prefixes the asset with `corbits:`.
 */
function detectFacilitator(envelope: unknown, entry: AcceptEntry): Facilitator {
  const env = envelope as Record<string, unknown> | null | undefined;
  const facilitator = (entry?.['extra'] as Record<string, unknown> | undefined)?.['facilitator'];
  if (typeof facilitator === 'string') {
    const f = facilitator.toLowerCase();
    if (f === 'payai') return 'payai';
    if (f === 'corbits') return 'corbits';
    if (f === 'coinbase') return 'coinbase';
  }
  const asset = asString(entry?.['asset']) ?? '';
  if (asset.startsWith('corbits:')) return 'corbits';
  if (env && typeof env['x402Version'] !== 'undefined') return 'coinbase';
  return 'unknown';
}

function parseEntry(envelope: unknown, entry: AcceptEntry): Requirement | null {
  const scheme = asString(entry['scheme']) ?? 'exact';
  const network = asString(entry['network']) ?? '';
  const asset = asString(entry['asset']) ?? '';
  const recipient =
    asString(entry['payTo']) ??
    asString(entry['recipient']) ??
    asString(entry['payee']) ??
    '';
  const amount =
    toBigint(entry['maxAmountRequired']) ??
    toBigint(entry['amount']) ??
    toBigint(entry['amountMicro']);

  if (!network || !asset || !recipient || amount === undefined) {
    return null;
  }

  const expiresAtRaw =
    asString(entry['expiresAt']) ?? asString(entry['expires_at']);
  const expiresAt = expiresAtRaw ? new Date(expiresAtRaw) : undefined;

  return {
    facilitator: detectFacilitator(envelope, entry),
    scheme,
    network,
    asset,
    amount,
    recipient,
    expiresAt: expiresAt && !Number.isNaN(expiresAt.getTime()) ? expiresAt : undefined,
    description: asString(entry['description']),
    raw: entry,
  };
}

/**
 * Parse a 402 body into a normalized `Requirement[]`.
 *
 * Returns an empty array when the body has no recognizable accepts list (the
 * caller should surface `ErrPaymentRequirementsInvalid` in that case).
 */
export function parsePaymentRequirements(body: unknown): Requirement[] {
  if (!body || typeof body !== 'object') return [];

  const obj = body as Record<string, unknown>;

  // Coinbase / PayAI / most facilitators: `{ accepts: [...] }`.
  let entries: AcceptEntry[] | null = null;
  if (Array.isArray(obj['accepts'])) {
    entries = obj['accepts'] as AcceptEntry[];
  } else if (Array.isArray(obj['paymentRequirements'])) {
    entries = obj['paymentRequirements'] as AcceptEntry[];
  } else if (Array.isArray(obj['requirements'])) {
    entries = obj['requirements'] as AcceptEntry[];
  }

  if (!entries) return [];
  const out: Requirement[] = [];
  for (const entry of entries) {
    if (entry && typeof entry === 'object') {
      const r = parseEntry(obj, entry);
      if (r) out.push(r);
    }
  }
  return out;
}

File: packages/sdk-py/rein/x402/parser.py
"""Parse 402 payment requirements from Coinbase / PayAI / Corbits facilitators.

Mirrors ``packages/sdk-ts/src/x402/parser.ts``. Anything that doesn't match
emerges as ``facilitator='unknown'`` so the caller can decide whether to
surface or reject.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal, Optional

Facilitator = Literal["coinbase", "payai", "corbits", "unknown"]


@dataclass(frozen=True)
class Requirement:
    facilitator: str
    scheme: str
    network: str
    asset: str
    amount: int
    recipient: str
    expires_at: Optional[datetime] = None
    description: Optional[str] = None
    raw: Any = None


def _to_int(v: Any) -> Optional[int]:
    if isinstance(v, int) and not isinstance(v, bool):
        return v
    if isinstance(v, float):
        return int(v)
    if isinstance(v, str) and v.isdigit():
        return int(v)
    return None


def _detect_facilitator(envelope: dict[str, Any], entry: dict[str, Any]) -> str:
    extra = entry.get("extra")
    if isinstance(extra, dict):
        f = extra.get("facilitator")
        if isinstance(f, str):
            f_low = f.lower()
            if f_low in {"payai", "corbits", "coinbase"}:
                return f_low

    asset = entry.get("asset", "")
    if isinstance(asset, str) and asset.startswith("corbits:"):
        return "corbits"

    if "x402Version" in envelope:
        return "coinbase"

    return "unknown"


def _parse_entry(envelope: dict[str, Any], entry: dict[str, Any]) -> Optional[Requirement]:
    scheme = entry.get("scheme") if isinstance(entry.get("scheme"), str) else "exact"
    network = entry.get("network") if isinstance(entry.get("network"), str) else None
    asset = entry.get("asset") if isinstance(entry.get("asset"), str) else None
    recipient = (
        entry.get("payTo")
        if isinstance(entry.get("payTo"), str)
        else entry.get("recipient")
        if isinstance(entry.get("recipient"), str)
        else entry.get("payee")
        if isinstance(entry.get("payee"), str)
        else None
    )
    amount = (
        _to_int(entry.get("maxAmountRequired"))
        if entry.get("maxAmountRequired") is not None
        else _to_int(entry.get("amount"))
        if entry.get("amount") is not None
        else _to_int(entry.get("amountMicro"))
    )

    if not network or not asset or not recipient or amount is None:
        return None

    expires_at_raw = entry.get("expiresAt") or entry.get("expires_at")
    expires_at: Optional[datetime] = None
    if isinstance(expires_at_raw, str):
        try:
            # Accept ISO 8601 with 'Z' suffix (Python 3.11+ handles this).
            expires_at = datetime.fromisoformat(expires_at_raw.replace("Z", "+00:00"))
        except ValueError:
            expires_at = None

    description = entry.get("description") if isinstance(entry.get("description"), str) else None

    return Requirement(
        facilitator=_detect_facilitator(envelope, entry),
        scheme=scheme or "exact",
        network=network,
        asset=asset,
        amount=amount,
        recipient=recipient,
        expires_at=expires_at,
        description=description,
        raw=entry,
    )


def parse_payment_requirements(body: Any) -> list[Requirement]:
    """Normalize a 402 body into a list of ``Requirement``s."""
    if not isinstance(body, dict):
        return []

    entries: Optional[list[Any]] = None
    if isinstance(body.get("accepts"), list):
        entries = body["accepts"]
    elif isinstance(body.get("paymentRequirements"), list):
        entries = body["paymentRequirements"]
    elif isinstance(body.get("requirements"), list):
        entries = body["requirements"]

    if entries is None:
        return []

    out: list[Requirement] = []
    for entry in entries:
        if isinstance(entry, dict):
            r = _parse_entry(body, entry)
            if r is not None:
                out.append(r)
    return out
